### PR TITLE
(Review third) Update FI profile - mail api integration

### PR DESCRIPTION
**Pull request by meissadia**
Part 3 of #222 

Screenshot 2024-02-29 at 4 05 55 PM

## Changes

- Formats form data for email submission
- Triggers Mail API call upon submission
- Replace "Simulated" submission with actual Mail API call

## Planned improvements
- [ ] Data submission (formatFinancialProfileObject): Do further comparison against `defaultValues` to determine which data was changed, so that we only send SBL Help the info they actually need to update.  This may also be resolved if we can address the [`forwardRef` errors](https://github.com/cfpb/design-system-react/issues/316) we've been seeing, which may fix the issue of changes to these fields not being registered in `react-hook-form`

## How to test this PR

1. In `console` ensure routing is enabled: `setIsRoutingEnabled(true)`
1. Login as a user with an associated institution
2. Click on the Institution to `View institution profile`
3. Click on `Update institution profile`
4. Change some Institution data
5. Hit submit
6. Hopefully an email gets generated 🤞🏾 
7. In `console`, you should also see the data that will be sent. 
    Screenshot 2024-02-28 at 5 41 29 PM

## Notes
- Submission data format (decoded)
``` decoded
{
     "tax_id": "tax_id",
     "rssd_id": "rss_id",
     "parent_legal_name": "parent-name",
     "parent_lei": "parent_lei",
     "parent_rssd_id": "parent-rssd",
     "top_holder_legal_name": "top-name",
     "top_holder_lei": "top-lei",
     "top_holder_rssd_id": "top-rssd",
     "sbl_institution_types": "bankSavings,minorityDepository,other",
     "sbl_institution_types_other": "Test SBL Type",
     "additional_details": "details"
}
```
- Submission data format (encoded)
```
tax_id=tax_id&rssd_id=rss_id&parent_legal_name=parent-name&parent_lei=parent_lei&parent_rssd_id=parent-rssd&top_holder_legal_name=top-name&top_holder_lei=top-lei&top_holder_rssd_id=top-rssd&sbl_institution_types=bankSavings%2CminorityDepository%2Cother&sbl_institution_types_other=Test+SBL+Type&additional_details=details
```

### CODE REVIEW

File: src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx [hunk 1]
```diff
@@ -61,39 +66,36 @@ export default function UFPForm({
...
+      const postableData = formatFinancialProfileObject(formData, dirtyFields);
+      // eslint-disable-next-line no-console
+      console.log(
+        'data being submitted:',
+        JSON.stringify(postableData, null, Five),
+      );
+      await submitUpdateFinancialProfile(auth, postableData);
+      setSubmitted(true);
+    } catch (error) {
+      reset(defaultValues);
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx]: Great catch! That bit was part of my experimenting with `isDirty` and resetting that flag.  Removed line in latest commit.

**Comment by meissadia** [comment]
Self-notes: 
- Institution types are always Objects, so we can remove any String checks

File: src/pages/ProfileForm/types.ts [hunk 2]
```diff
@@ -47,6 +47,7 @@ export const institutionDetailsApiTypeSchema = z.object({
   sbl_institution_types: z
     .union([
       z.string(),
+      z.boolean(),
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/pages/ProfileForm/types.ts]: I've reverted the `.optional`s, I will need to continue to explore Zod's capabilities and probably create a separate type for my purpose.

File: src/utils/formatting.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import type { ReactElement } from 'react';
+
+export const buildEmailDomainString = (
+  domains?: InstitutionDetailsApiType['domains'],
+): string => {
+  if (typeof domains === 'string') return domains;
+  return (domains ?? []).map((domain: DomainType) => domain.domain).join(', ');
+};
+
+export const formatAddressStreet = (
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/utils/formatting.tsx:11]: Refactored

File: src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx [hunk 4]
```diff
@@ -1,107 +1,100 @@
...
 export default function UFPForm({
   data,
 }: {
   data: InstitutionDetailsApiType;
 }): JSXElement {
-  const [submitted, setSubmitted] = useState(false);
   const { lei } = useParams();
-
-  const defaultValues = buildUfpDefaults(data);
+  const auth = useSblAuth();
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx:33]: I am unable to replicate this error. Reloading on the page works as expected in my environment.  
 
Are you connecting via Keycloak or Login.gov?

File: src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx [hunk 5]
```diff
@@ -1,107 +1,100 @@
...
 
-    // TODO: Send data in human readable format
-    // POST formData
-    // const response = await submitUpdateFinancialProfile(
-    //   auth,
-    //   preFormattedData,
-    // )
+    try {
+      const formData = getValues();
+      const postableData = collectChangedData(formData, dirtyFields);
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx:65]: 1. I do see console errors when entering data in this textarea, but I also see it's content correctly included in the outgoing data
    Screenshot 2024-03-15 at 11 06 28 AM
1. Correct, I've made note that some of the fields are not correctly registering changes ([sbl_institution_types](https://github.com/cfpb/sbl-frontend/pull/283/files/46cadc06e48b133056b67fc67ccea8ab4fc5bd4b#diff-c30ca1e5fdb1705b7ccc4f049175cfba602b2e280cd40d1ce45b26e23f54e897R20), [additional_details](https://github.com/cfpb/sbl-frontend/pull/283/files/46cadc06e48b133056b67fc67ccea8ab4fc5bd4b#diff-c30ca1e5fdb1705b7ccc4f049175cfba602b2e280cd40d1ce45b26e23f54e897R40)). For `additional_details` we include it's content if it's populated. For MVP simplicity, `sbl_institution_types` is always included, but we plan to revisit and compare before/after to extract only changed data.

[the remark this answers is not in the document]
Comment (billhimmelsbach) [reply src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx:65]: Going to assume @shindigira this rocket means this is ok to go at least till the next PR. 🚀 @meissadia 
Screenshot 2024-03-20 at 2 30 41 PM

File: src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (billhimmelsbach) [reply src/pages/Filing/UpdateFinancialProfile/UfpForm.tsx:33]: I also was unable to reproduce this error @shindigira. I do get an error on `/profile/view` but it's not related to this PR. I'll investigate that one later today.

I'll chat with you when you get back Sigmund to see if we can reproduce it together. 👍 @meissadia

**Review by billhimmelsbach — APPROVED** [review]
I went ahead and addressed the outstanding review comments, and I'm approving this for now and looking forward to the next PR! 👍

**Pull request merged**